### PR TITLE
Hw28 wed

### PR DESCRIPTION
**Pull request by fyarkiy**

### CODE REVIEW

File: src/main/java/internet/store/controllers/InjectDataController.java [hunk 1]
```diff
@@ -25,18 +27,11 @@
...
-        User mandy = new User("Mandy", "Mandy", "1234321");
-        userService.create(mandy);
-        shoppingCartService.create(new ShoppingCart(mandy.getUserId()));
-        User joanne = new User("Joanne", "Joe", "1234321");
-        userService.create(joanne);
-        shoppingCartService.create(new ShoppingCart(joanne.getUserId()));
+        User admin = new User("Admin", "admin", "1");
+        admin.setRoles(Set.of(Role.of("ADMIN")));
+        userService.create(admin);
+        shoppingCartService.create(new ShoppingCart(admin.getUserId()));
```
Comment (AnaLanda) [inline src/main/java/internet/store/controllers/InjectDataController.java]: Our home task says that only USERs can access shopping cart operations, so maybe there's no need to create a shopping cart for the ADMIN?

File: src/main/webapp/web.xml [hunk 2]
```diff
@@ -166,6 +166,24 @@
...
+        <url-pattern>/logout</url-pattern>
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>accessDenied</servlet-name>
+        <servlet-class>internet.store.controllers.usercontroller.LogoutController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>accessDenied</servlet-name>
+        <url-pattern>/user/accessDenied</url-pattern>
```
Comment (AnaLanda) [inline src/main/webapp/web.xml]:
```suggestion
        <url-pattern>/user/access-denied</url-pattern>
```

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    public void init(FilterConfig filterConfig) throws ServletException {
+        protectedUrls.put("/order/all", Set.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/product/create", Set.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/product/manage", Set.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/user/all", Set.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/user/delete", Set.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/product/all", Set.of(Role.RoleName.USER));
+        protectedUrls.put("/shopping-cart/product", Set.of(Role.RoleName.USER));
+        protectedUrls.put("/user/order", Set.of(Role.RoleName.USER));
+    }
```
Comment (AnaLanda) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]: What about these URLs: 
`/product/delete`
`/shopping-cart/product/add`
`/shopping-cart/product/remove`
`/order/create`
`/order/delete`
?

**Review by AnaLanda — COMMENTED** [review]
Well done! But I think you can add more URLs to protectedUrls 😉

File: src/main/webapp/web.xml (hunk 2 above)
Comment (fyarkiy) [reply src/main/webapp/web.xml]: Noticed accessDenied in Bohdan's video but your solution is nicer

File: src/main/java/internet/store/controllers/InjectDataController.java (hunk 1 above)
Comment (fyarkiy) [reply src/main/java/internet/store/controllers/InjectDataController.java]: make sense

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 3 above)
Comment (fyarkiy) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java]: Just Tested - you're right, from url I can reach to this menu even when access to the page with references to those methods is blocked...

File: src/main/java/internet/store/model/Role.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package internet.store.model;
+
+public class Role {
+    private Long id;
+    private RoleName roleName;
+
+    public Role(RoleName roleName) {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/model/Role.java]:
```suggestion
    private Role(RoleName roleName) {
```

File: src/main/webapp/WEB-INF/views/user/access-denied.jsp [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Title</title>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/user/access-denied.jsp]: add informative title

File: src/main/webapp/web.xml [hunk 6]
```diff
@@ -166,6 +166,24 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>accessDenied</servlet-name>
+        <servlet-class>internet.store.controllers.usercontroller.LogoutController</servlet-class>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>accessDenied</servlet-name>
+        <url-pattern>/user/access-denied</url-pattern>
+    </servlet-mapping>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]: do we have this controller?

File: src/main/webapp/web.xml [hunk 7]
```diff
@@ -174,4 +192,13 @@
         <filter-name>authenticationFilter</filter-name>
         <url-pattern>/*</url-pattern>
     </filter-mapping>
+
+    <filter>
+        <filter-name>autharisationFilter</filter-name>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <filter-name>authorizationFilter</filter-name>
```

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AuthorisationFilter implements Filter {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java:20]:
```suggestion
public class AuthorizationFilter implements Filter {
```

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        } else {
+            req.getRequestDispatcher("/WEB-INF/views/user/access-denied.jsp").forward(req, resp);
+        }
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    private boolean isAthourised(User user, Set<Role.RoleName> authorisedRoles) {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]:
```suggestion
    private boolean isAuthourized(User user, Set<Role.RoleName> authorisedRoles) {
```

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String requestedUrl = req.getServletPath();
+
+        if (protectedUrls.get(requestedUrl) == null) {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]: let's try to put both cases in one `if`
and use `containsKey()` for this condition

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            chain.doFilter(req, resp);
+            return;
+        }
+
+        Long userId = (Long) req.getSession().getAttribute("user_id");
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]:
```suggestion
        Long userId = (Long) req.getSession().getAttribute(USER_ID);
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 8 above)
Comment (fyarkiy) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java:20]: How critical is it? With "z" - it's American English, with "s" - British. May I leave British one?

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 10 above)
Comment (fyarkiy) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java]: looks a bit harder to read but less lines - will do changes!

File: src/main/webapp/web.xml (hunk 7 above)
Comment (fyarkiy) [reply src/main/webapp/web.xml]: the same, pleaaase? I don't like "zee":(...

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 9 above)
Comment (fyarkiy) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java]: May I stay with British, please?

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 8 above)
Comment (KseniiaMakarova) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java:20]: ok)

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        } else {
+            req.getRequestDispatcher("/WEB-INF/views/user/access-denied.jsp").forward(req, resp);
+        }
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    private boolean isAthourised(User user, Set<Role.RoleName> authorisedRoles) {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]:
```suggestion
    private boolean isAuthourised(User user, Set<Role.RoleName> authorisedRoles) {
```

File: src/main/webapp/web.xml [hunk 13]
```diff
@@ -174,4 +183,13 @@
...
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <filter-name>authorisationFilter</filter-name>
```

File: src/main/webapp/web.xml [hunk 14]
```diff
@@ -174,4 +183,13 @@
...
+        <filter-class>internet.store.web.filters.AuthorisationFilter</filter-class>
+    </filter>
+    <filter-mapping>
+        <filter-name>autharisationFilter</filter-name>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml]:
```suggestion
        <filter-name>authorisationFilter</filter-name>
```

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String requestedUrl = req.getServletPath();
+
+        if (!protectedUrls.containsKey(requestedUrl)
+                || isAthourised(userService.get((Long) req.getSession().getAttribute(USER_ID)),
```
Comment (KseniiaMakarova) [inline src/main/java/internet/store/web/filters/AuthorisationFilter.java]: to make this more readable, we can move `(Long) req.getSession().getAttribute(USER_ID)` to local variable

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/internet/store/web/filters/AuthorisationFilter.java (hunk 9 above)
Comment (fyarkiy) [reply src/main/java/internet/store/web/filters/AuthorisationFilter.java]: Thx!!!

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**